pkg/providers/amazon/autoscaling/prometheus.go

```diff
@@ -0,0 +1,128 @@
+// Copyright © 2018 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoscaling
+
+import (
+	"github.com/sirupsen/logrus"
+
+	"github.com/banzaicloud/pipeline/pkg/providers/amazon"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	timers map[string]*prometheus.Timer
+
+	ec2InstanceStartupDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace: "pipeline",
+		Name:      "ec2_instance_startup_duration_seconds",
+		Help:      "EC2 instance startup duration in seconds",
+	},
+		[]string{"provider", "region", "zone", "type", "price_type"},
+	)
+	ec2SpotInstanceFulfillmentDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace: "pipeline",
+		Name:      "ec2_spot_instance_fulfillment_duration_seconds",
+		Help:      "EC2 spot instance fulfillment duration in seconds",
+	},
+		[]string{"provider", "region", "zone", "type"},
+	)
+)
+
+func init() {
+	timers = make(map[string]*prometheus.Timer)
+	prometheus.MustRegister(ec2InstanceStartupDuration)
+	prometheus.MustRegister(ec2SpotInstanceFulfillmentDuration)
+}
+
+// StartMetricTimer starts a prometheus metric timer for a node instance
+func (m *Manager) StartMetricTimer(instance *Instance) *prometheus.Timer {
+	var region, availabilityZone, instanceType, priceType string
+
+	if !m.metricsEnabled {
+		return nil
+	}
+
+	key := *instance.InstanceId
+	if timers[key] != nil {
+		return timers[key]
+	}
+
+	m.logger.WithField("instance-id", *instance.InstanceId).Debug("start metric timer")
+
+	lc, err := instance.GetLaunchConfiguration()
```

Comment:
instead of relying on `LaunchConfigurations` to determine if the instance is spot or not, please consider using `DescribeInstances` instead and parsing `instanceLifecycle` from the response. There is only a slim chance, but `LaunchConfigs` may change or be deleted after the instance is started.

Reply:
Sure, coming right up! 👍